Annotation example (+ stack component update) (#813)
zenml-io/zenml: examples/label_studio_annotation/README.md

# 🏷 Continuous Annotation with Label Studio

Data annotation / labeling is a core part of MLOps that is frequently left out
of the conversation. ZenML offers a way to build continuous annotation (combining
training and annotation into a loop) with Label Studio. This uses a combination
of user-defined steps as well as some built-in steps that ZenML provides.

> ## Basic guide to running this example
> 
> This example requires a few more steps than usual, given that it requires a
> cloud artifact store. Instructions for Azure, AWS S3 and GCP can be found in the
> following guide.
> 
> ## 🗺 Overview
> 
> NOTE: This example currently only runs with a cloud artifact store. [This guide
> by Label Studio](https://labelstud.io/guide/storage.html) contains more details
> if you have trouble understanding the instructions below.

Reviewer: This feels a bit redundant, would propose to remove the second paragraph and move the first one down:

```suggestion
## 🗺 Overview

This example requires a few more steps than usual, given that it requires a
cloud artifact store. Instructions for Azure, AWS S3 and GCP can be found in the
following guide.
```

> For a full video walkthrough on how to run this email, please check out [our

Reviewer 2: ```suggestion
For a full video walkthrough on how to run this example, please check out [our
```

community hour demo](https://www.youtube.com/watch?v=bLFGnoApWeU) in which we
show how to run the various pipelines, in what order, all using an AWS stack.

# 🖥 Run the example

## 👣 Step-by-Step

> 

Reviewer: I'd suggest to add a header "Infrastructure Setup" here under which you group the three sections for AWS, Azure, GCP as subheaders and also add a short comment in the beginning clarifying again that you can use any of the three cloud providers. 

(When I first saw this it felt like you need to do all three setups since "Step-by-Step" sounds to me like you need to follow the subsequent steps one-by-one.)

### 🥞 Set up your stack for Microsoft Azure

In order to run this example, you need to install and initialize ZenML and Label
Studio.

```shell
zenml integration install label_studio pytorch azure

# pull example
zenml example pull label_studio_annotation
cd zenml_examples/label_studio_annotation

# Initialize ZenML repo
zenml init

> ```

Reviewer: This is essentially the same for all three cloud providers, only `zenml integration install azure` is provider-specific here. I would recommend to move the rest into a separate section "ZenML Setup" or similar to remove code duplication and make the README a bit shorter

Author: Yep I moved that now.

zenml stack update <ANNOTATION_STACK_NAME> -a <YOUR_CLOUD_ARTIFACT_STORE>
```

You will next need to obtain your Label Studio API key. This will give you access to the web annotation interface.

```shell
# start a temporary / one-off label-studio instance to get your API key
label-studio start -p 8094
```

First order of business is to sign up if you haven't done so already, using an email and a password. Then visit [http://localhost:8094/](http://localhost:8094/) to log in, and then visit [http://localhost:8094/user/account](http://localhost:8094/user/account) and get your Label Studio API key (from the upper right hand corner). You will need it for the next step. `Ctrl-c` out of the Label Studio server that is running on the terminal.

```shell
zenml secret register <LABEL_STUDIO_SECRET_NAME> --api_key="<YOUR_API_KEY>"

zenml annotator register <YOUR_LABEL_STUDIO_ANNOTATOR> --flavor label_studio --authentication_secret="<LABEL_STUDIO_SECRET_NAME>"

zenml stack update <ANNOTATION_STACK_NAME> -an <YOUR_LABEL_STUDIO_ANNOTATOR>

zenml stack up

> ```

Reviewer: This is also not cloud provider-specific. Similar to above, I would suggest to move this to a separate section "Label Studio Setup" or similar instead of having it duplicated 3x.

Author: Also has been moved, but the stack setup steps can't be moved elsewhere unfortunately as there are some order dependencies that are a bit fiddly.

zenml stack update <YOUR_AWS_ZENML_STACK_NAME> -x <YOUR_SECRETS_MANAGER>

# copy down the credentials output from this command
aws sts get-session-token --profile=<YOUR_AWS_CONFIG_PROFILE_NAME>
```

Use the credentials output from the command to get your `<LABEL_STUDIO_ACCESS_KEY_ID>`, `<LABEL_STUDIO_SECRET_ACCESS_KEY>`, and `<LABEL_STUDIO_AWS_SESSION_TOKEN>` which we will use to register a secret.

```shell
# this secret must be named 'aws_label_studio'
zenml secret register aws_label_studio --schema=aws_label_studio --aws_access_key_id="<LABEL_STUDIO_ACCESS_KEY_ID>" --aws_secret_access_key="<LABEL_STUDIO_SECRET_ACCESS_KEY>" --aws_session_token="<LABEL_STUDIO_AWS_SESSION_TOKEN>"

# use your standard access key id and secret access key from ~/.aws/credentials here
zenml secret register <YOUR_AWS_SECRET_NAME> --schema=aws_label_studio--aws_access_key_id="<YOUR_ACCESS_KEY_ID>" --aws_secret_access_key="<YOUR_SECRET_ACCESS_KEY>"

zenml artifact-store register <YOUR_CLOUD_ARTIFACT_STORE> --flavor=s3 --path=s3://<YOUR_S3_BUCKET_NAME> --authentication_secret="<YOUR_AWS_SECRET_NAME>"

zenml stack update <YOUR_AWS_ZENML_STACK_NAME> -a <YOUR_CLOUD_ARTIFACT_STORE>
```

> You will next need to obtain your Label Studio API key. This will give you access to the web annotation interface.

Reviewer 2: Can we move this up ahead of the split into the three cloud providers to reduce the duplication?

```shell
# start a temporary / one-off label-studio instance to get your API key
label-studio start -p 8094
```

First order of business is to sign up if you haven't done so already, using an email and a password. Then visit http://localhost:8094/ to log in, and then visit http://localhost:8094/user/account and get your Label Studio API key (from the upper right hand corner). You will need it for the next step. Ctrl-c out of the Label Studio server that is running on the terminal.

```shell
zenml secret register <LABEL_STUDIO_SECRET_NAME> --api_key="<YOUR_API_KEY>"

zenml annotator register <YOUR_LABEL_STUDIO_ANNOTATOR> --flavor label_studio --authentication_secret="<LABEL_STUDIO_SECRET_NAME>"

zenml stack update <YOUR_AWS_ZENML_STACK_NAME> -an <YOUR_LABEL_STUDIO_ANNOTATOR>

zenml stack up
```

### ▶️ Run the Code

> This will initialize the daemon server which Label Studio requires, albeit running on a default port of 8093.

Reviewer: This line should be moved after `zenml stack up` I guess?

> This will initialize the daemon server which Label Studio requires, albeit running on a default port of 8093.

Reviewer 2: Should this sentence be part of `Run the Code`?

> Run the pipeline with:

Reviewer: This needs a bit more explanation, e.g.:

#### 1. Run the training pipeline
This does ...
You should see ...

#### 2. Run the inference pipeline
...
This creates new data automatically written to label studio backend blabla
...

#### 3. Open Label Studio and annotate the new data
Go to this URL, click here, do that, now you should see such and such
....

#### 4. Rerun the training data
This is what happens, trains on new data, ...

#### 5. Rerun the inference pipeline
...

#### (Optional) 6. Annotate new data in label studio again
...

```shell
python run.py aws --train
python run.py aws --inference
# Now do some annotation using Label studio
python run.py aws --train --rerun
python run.py aws --inference --rerun
```

> ## CLI Commands for the Label Studio Integration

Reviewer: Instead of listing what commands there are I'd recommend to show it at the respective place in the previous section, e.g., if you follow the structure I proposed in the previous comment then you could under 2. do `zenml annotator dataset list` at the end to show that there is now something in label studio, and then under 3. have the `zenml annotator dataset ... stats` command to show that X images need to be labeled now and then run `zenml annotator dataset ... annotate` to open the UI and do the annotation. Screenshots of these steps (label studio UI, stats CLI output, ...) would also be nice to include here so users can understand the example without having to actually run it.

Once you've run the pipeline for the first time, you'll be able to use some of
the ZenML CLI commands to interact with your Label Studio annotations and the
dataset:

```shell
# the obvious ones to try
zenml annotator describe
zenml annotator dataset list
zenml annotator dataset <YOUR_DATASET_NAME> stats
```

### 🧽 Clean up

In order to clean up, delete any relevant cloud infrastructure that was created

> above (the S3 bucket is the main one, along with any permissions) and the

Reviewer 2: not just s3 but also azure/gcp + secrets manager right?

> remaining ZenML references.

Reviewer: Would be cool to have a short list of what exactly needs to be deleted and/or links to guides how to do it. @wjayesh  helped me with this for the k8s orchestrator example/blog, maybe we could copy-paste that to here (for AWS at least).

```shell
rm -rf zenml_examples
```

# 📜 Learn more

If you want to learn more about annotation in general or about how to use your
own annotation tool in ZenML
check out our [docs](https://docs.zenml.io/mlops-stacks/annotators).

> We also have extensive CLI docs for the
> [annotator](https://apidocs.zenml.io/latest/api_docs/annotators/) stack
> component.

Reviewer: ```suggestion
To explore other available CLI commands related to the annotator stack components, check out the
[CLI docs annotator section](https://apidocs.zenml.io/latest/api_docs/repository/#zenml.repository.Repository).
```
